demos/python_demos/object_detection_demo_ssd_async/object_detection_demo_ssd_async.py

```diff
@@ -15,239 +15,589 @@
+
+        self.requests = self.exec_net.requests
+        self.empty_requests = deque(self.requests)
+        self.completed_request_results = results if results is not None else []
+        self.callback_exceptions = caught_exceptions if caught_exceptions is not None else {}
+        self.event = threading.Event()
+
+    @staticmethod
+    def check_cpu_support(ie, net):
+        log.info('Check that all layers are supported...')
+        supported_layers = ie.query_network(net, 'CPU')
+        not_supported_layers = [l for l in net.layers.keys() if l not in supported_layers]
+        if len(not_supported_layers) != 0:
+            unsupported_info = '\n\t'.join('{} ({} with params {})'.format(layer_id,
+                                                                           net.layers[layer_id].type,
+                                                                           str(net.layers[layer_id].params))
+                                           for layer_id in not_supported_layers)
+            log.warning('Following layers are not supported '
+                        'by the CPU plugin:\n\t{}'.format(unsupported_info))
+            log.warning('Please try to specify cpu extensions library path.')
+            raise ValueError('Some of the layers are not supported.')
+
+    def unify_inputs(self, inputs):
+        if not isinstance(inputs, dict):
+            if not isinstance(inputs, (list, tuple)):
+                inputs = [inputs]
+            if len(inputs) == 1:
+                inputs_dict = {next(iter(self.net.input_info)): inputs[0]}
+            else:
+                raise ValueError
+        else:
+            inputs_dict = inputs
+        return inputs_dict
+
+    def preprocess(self, inputs):
+        meta = {}
+        return inputs, meta
+
+    def postprocess(self, outputs, meta):
+        return outputs
+
+    def inference_completion_callback(self, status, callback_args):
+        request, frame_id, frame_meta = callback_args
+        try:
+            if status != 0:
+                raise RuntimeError('Infer Request has returned status code {}'.format(status))
+            raw_outputs = {key: blob.buffer for key, blob in request.output_blobs.items()}
+            self.completed_request_results[frame_id] = (frame_meta, raw_outputs)
+            self.empty_requests.append(request)
+        except Exception as e:
+            self.callback_exceptions.append(e)
+        self.event.set()
+
+    def __call__(self, inputs, id, meta):
+        request = self.empty_requests.popleft()
+        inputs = self.unify_inputs(inputs)
+        inputs, preprocessing_meta = self.preprocess(inputs)
+        meta.update(preprocessing_meta)
+        request.set_completion_callback(py_callback=self.inference_completion_callback,
+                                        py_data=(request, id, meta))
+        self.event.clear()
```

Comment:
I don't think this is correct. Consider this sequence of events:

1. Main thread starts one InputRequest.
2. It completes in the background thread, setting the event.
3. Main thread starts the next InputRequest (this clears the event).
4. Main thread calls `await_any`. This ignores the first request's completion (because the event was cleared) and waits for the second request.

Reply:
See no issues here. We are calling `await_any` right after the check that all Infer Requests are busy. In your situation we won't call `await_any` at all, which is fine.

Reply:
OK, I see what you mean. It does seem safe (if a bit unobvious).